### PR TITLE
Add aten::masked_select and its variants

### PR DESCRIPTION
**Pull request by xytintel**
Task list:
- [x] masked_select
- [x] masked_select.out

Fixes: #629

### CODE REVIEW

**Comment by xytintel** [comment]
image
image

**Comment by dvrogozh** [comment]
@xytintel : add the following to PR description of commit message to automatically close related issue:
```
Fixes: #629
```

File: src/ATen/native/xpu/Indexing.cpp [hunk 1]
```diff
@@ -44,4 +45,53 @@ Tensor XPUNativeFunctions::index_select(
...
+      : c10::MaybeOwned<Tensor>::borrowed(mask);
+  auto self_temp = (self.dim() == 0)
+      ? c10::MaybeOwned<Tensor>::owned(self.unsqueeze(0))
+      : c10::MaybeOwned<Tensor>::borrowed(self);
+
+  // Cannot reassign to mask_temp and self_temp here! if they are
+  // owning and expand_outplace returns a borrow, the returned borrow
+  // would dangle.
+  auto mask_self_expanded = expand_outplace(*mask_temp, *self_temp);
+  at::index_out(
```
Comment (fengyuan14) [inline src/ATen/native/xpu/Indexing.cpp:72]: Any concern of the additional redispatch?
https://github.com/pytorch/pytorch/blob/63374dda6999a2541db447ef9984ac7c2e82ea3c/aten/src/ATen/native/cuda/IndexKernel.cpp#L43

Comment (xytintel) [reply src/ATen/native/xpu/Indexing.cpp]: Fixed

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**